sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -345,6 +345,8 @@ class Analyzer(override val catalogManager: CatalogManager)
     override def apply(plan: LogicalPlan): LogicalPlan = plan.resolveOperatorsUp {
       case p: LogicalPlan => p.transformExpressionsUp {
         case a @ Add(l, r, f) if a.childrenResolved => (l.dataType, r.dataType) match {
+          case (DateType, DayTimeIntervalType) => TimeAdd(Cast(l, TimestampType), r)
+          case (DayTimeIntervalType, DateType) => TimeAdd(Cast(r, TimestampType), l)
```

Comment:
I doubt this. For old interval types, `Date + Interval` always returns Date:
```
case (DateType, CalendarIntervalType) => DateAddInterval(l, r, ansiEnabled = f)
case (CalendarIntervalType, DateType) => DateAddInterval(r, l, ansiEnabled = f)
```

I think we should do this for new interval types as well, and fail at runtime if the interval has hour, minute, second fields.